### ISSUE TITLE
[explicit-function-return-type] Is type inference not allowed?

### ISSUE THREAD

**Issue by KhakiMosaic** [typescript-eslint/typescript-eslint#434]
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/explicit-function-return-type": "error"
  }
}
```

```TS
const fn1: (i: number) => string = function(i) { return i.toString(); }; // Error
const fn2: (i: number) => string = i => i.toString(); //Error
```

**Expected Result**
Expect no error because these functions are inferred to return string.

**Actual Result**
Error. (Missing return type on function.)

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `1.6.0` |
| `@typescript-eslint/parser`        | `1.6.0` |
| `TypeScript`                       | `3.4.3` |
| `ESLint`                           | `5.16.0` |
| `node`                             | `10.15.3` |
| `npm`                              | `6.4.1` |

**Comment by bradzacher** [comment]
Check out the options for the rule: https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/explicit-function-return-type.md#options

I believe you'll want `allowTypedFunctionExpressions`

**Comment by hanneswidrig** [comment]
This might be a dumb question but I am trying to understand how I use these rules for [explicit-function-return-type] inside my React Typescript application on Create-React-App 3.0.

```
{
  "plugins": ["react", "react-hooks", "@typescript-eslint"],
  "extends": [
    "plugin:@typescript-eslint/recommended",
    "prettier",
    "prettier/@typescript-eslint"
  ],
  "rules": {
    "react-hooks/rules-of-hooks": "error",
    "react-hooks/exhaustive-deps": "warn",
    "allowExpressions ": true,
    "allowTypedFunctionExpressions": true
  },
  "parser": "@typescript-eslint/parser",
  "globals": {
    "Atomics": "readonly",
    "SharedArrayBuffer": "readonly"
  },
  "env": {
    "browser": true,
    "es6": true
  },
  "parserOptions": {
    "project": "./tsconfig.json",
    "ecmaFeatures": {
      "jsx": true
    },
    "ecmaVersion": 2018,
    "sourceType": "module"
  }
}
```

This is my .eslintrc.json file. I am wondering how I add those rules like `AllowExpressions`. 

`"@typescript-eslint/explicit-function-return-type": 0,` Adding this to the rules made it go away but I am not sure if this is completely turning off all rules. I still want to require function return type when it's not inferred.

![image](https://user-images.githubusercontent.com/3399497/56745480-6f0dea80-6748-11e9-899e-6bab22120d55.png)

**Comment by bradzacher** [comment]
@hanneswidrig

Check out the eslint docs for configuring rules.
https://eslint.org/docs/user-guide/configuring#using-configuration-files

The available options for this rule are listed here
https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/explicit-function-return-type.md#options

Let me know if you need more help.

**Comment by hanneswidrig** [comment]
@bradzacher I just figured it out, thanks for the resources though. 

For anyone working with React and wondering about this, see the example below.
![image](https://user-images.githubusercontent.com/3399497/56745919-41757100-6749-11e9-8654-109ccb6952d4.png)
